### DISCUSSION TITLE
Camus running on Raspberry pi 4 under Ubuntu server 20.10

### DISCUSSION THREAD

**Discussion by michelmimi** [mrgnr/camus#25] [General]
Camus is quite easy to use, however, I encounter some difficulties with its setup on Ubuntu server 20.10 running on a Raspberry pi 4.  

- Installation with the command line: **$ sudo snap install camus** is a breeze.

- Creating the room with the request: **localhost:5000** is straightforward

But how do other clients connect to the same room from their browser? Some attempts on the same network segment were unsuccessful. I probably missed something in the setup. Why isn't it possible to create the room remotely with the following command: <camus_server_IP>:5000 ?

**Comment by mrgnr** [comment]
Hi @michelmimi, this is an excellent question. There seem to be two main issues to solve here:
1. Camus is (currently) [hard-coded](https://github.com/mrgnr/camus/blob/ff3c125a3606db7baef08c9dee434ff50277d7d0/bin/camus#L1) to listen on `127.0.0.1:5000`, so port 5000 is not accessible to other machines on the network.
2. To access the camera & mic, web browsers require a ["secure context"](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia#privacy_and_security), which means HTTPS is necessary if using an address other than `http://localhost`.

To solve the first issue, you can run a reverse proxy like Nginx or Apache to forward traffic from a public port to Camus at `127.0.0.1:5000`. You can see the [template here](https://github.com/mrgnr/terraform-camus/blob/71512299ecc777b7a774b699974b9173f447c05b/nginx.conf) as a reference for the Nginx configuration. (Note that `${domain}` in that file should be substituted with the hostname or static IP address of your Pi.)

To enable HTTPS, you'll need an SSL certificate. For a public server, you can use [Certbot](https://certbot.eff.org/) to get one from Let's Encrypt. For a non-public server (i.e. if you want to access Camus on your local network only), you will have to [generate self-signed certificates](https://www.digitalocean.com/community/tutorials/openssl-essentials-working-with-ssl-certificates-private-keys-and-csrs#generating-ssl-certificates). (Note that using a self-signed certificate will cause your browser to show a security warning, since the certificate isn't signed by a trusted third-party. To get rid of this warning, you'll have to go through some extra work to create your own certificate authority (CA), use it to sign your server certificate, then add the public CA certificate to any web browser that needs to access your Camus server.)

Let me know if this helps, and if not I can try to give some more tips or point you to additional resources.

P.s. If anyone else reading this has struggled or succeeded with setting up a Camus server on their local network, feel free to share feedback or tips!

**Reply by michelmimi** [reply]
Hello, sorry for my late reply. Thank you developing this video chat application that seems to give good results **even with a Raspberry pi 4 running as workstation with both video and audio**. It was in this case tested between two participants. Compared to Jitsi (also WebRTC based), Camus doesn't need a very powerful server. Camus was successfully tested with four persons within our LUG but in this later case with 4 PCs. I finally was able to build [my own server / Heroku](https://docs.camus.chat/en/latest/deployment.html) . But the neat thing appears to be the tutorial you just created to build his own server on a Raspberry pi 4. That's great!

**Comment by mrgnr** [comment]
I tried setting this up myself and wrote a tutorial: [How to deploy Camus on Raspberry Pi with Ubuntu Server](https://gist.github.com/mrgnr/76ed5e1e3d0a2cd192832e701d0edd25). Comments appreciated :)

cc @michelmimi

**Reply by michelmimi** [reply]
Thank you for this tutorial.  OK, I'll get rid of my Ubuntu 20.10 installation, will follow your tutorial by installing the Ubuntu LTS 20.04 version and will let you know the result.

**Reply by mrgnr** [reply]
@michelmimi Thanks for trying this out!

Hmm, I think the 502 error might be because Nginx couldn't find the upstream. Make sure that the value for `server_name` in `/etc/nginx/conf.d/camus.conf` matches the hostname and try restarting Nginx. In the other case where you got the "Welcome to Nginx!" page with the server IP, it's probably because Nginx is returning the default page since the IP isn't matching anything found in `camus.conf`.

Thanks for the comment about updating `/etc/hosts`, I will add it to the tutorial :+1: . I was relying on my home router's DNS, but that might not work for everyone.

The output from `curl` looks fine, the `/` path should redirect to `/chat`. If you go to `http://camus` in your browser, it will automatically redirect to `http://camus/chat`.

**Reply by michelmimi** [reply]
@mrgnr 

May I point out that I updated `/etc/hosts` both on my workstation and "camus" server by adding the line:
`<my_camus_server_ip_address>    camus` 

If someone would want to run an Ubuntu 20.04.2 **headless server**, the tutorial may perhaps add following time zone configuration:

`$ timedatectl list-timezones  # to list all time zones`

`$ sudo timedatectl set-timezone < your_time_zone >   # to define own time zone`

**Camus is running now flawless on my Raspberry pi 4 with Ubuntu server 20.04.2**.

**Reply by mrgnr** [reply]
I'm glad to hear everything is working for you! I finally got around to merging the tutorial into the docs, available [here](https://docs.camus.chat/en/latest/tutorials/rpi-deployment.html).
